<div align="right">
<a href="http://nbviewer.ipython.org/github/CSP-UHS/Hermons_Book_of_Python/blob/master/Chapter00.ipynb" align="right"><h2>Table of Contents</h2></a>
</div>

# Chapter 7: Text Formatting

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Here is a quick table for reference when doing text formatting. For a detailed explanation of how text formatting works so you can go back and clean up the formatting in all of your programs, keep reading this incredibly exciting chapter!

<center><img src="http://www.hermonswebsites.com/PyNotebooks/Python/07/formatting.png"></center>

## Decimal Numbers

Try running the following program, which prints out several random numbers. The output is left justified and numbers look terrible.

In [4]:
import random
 
for i in range(10):
    x = random.randrange(20)
    print(x)

3
9
18
0
4
4
4
17
14
6


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We can use string formatting to make the list of numbers look better by right-justifying them. The first step is to use the format command on the string as seen below. See how the string ends with .format(x). In future chapters we will talk about classes, instances(objects) and methods. All strings are actually instances of a class named String. That class has methods that can be called. One of them is format.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The format function will not print out the curly braces { } but instead replaces them with the value in x. The output (below) looks just like what we had before so obviously we are supposed to put something into the curly braces.

In [6]:
import random
 
for i in range(10):
    x = random.randrange(20)
    print("{}".format(x) )

0
4
9
2
2
0
13
12
9
7


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;To right justify, we add more information about how to format the number between the curly braces { }:

In [7]:
import random
 
for i in range(10):
    x = random.randrange(20)
    print("{:2}".format(x) )

12
18
 6
 2
 4
16
 0
11
 9
 3


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is better; we have right justified numbers! But how does it work? The :2 that we added isn't exactly obvious.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Here's the breakdown: The { } tells the computer we are going to format a number. After the : inside the curly braces will be formatting information. In this case we give it a 2 to specify a field width of two characters. The field width value tells the computer to try to fit the number into a field two characters wide. By default, it will try to right-justify numbers and left-justify text.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Even better, the program no longer needs to call str( ) to convert the number to a string. Leave the string conversions out. What if you had large numbers? Let's make bigger random numbers below.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We added a comma after the field width specifier, and now our numbers have commas. That comma must go after the field width specifier, not before. Commas are included in calculating the field width. For example, 1,024 has a field width of 5, not 4.

In [9]:
import random

for i in range(10):
    x = random.randrange(100000)
    print("{:6,}".format(x) )

79,446
65,460
20,376
36,919
21,323
48,935
79,647
52,971
33,605
78,656


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We can print multiple values, and combine the values with text. Run the code below.

In [10]:
x = 5
y = 66
z = 777
print("A - '{}' B - '{}' C - '{}'".format(x, y, z))

A - '5' B - '66' C - '777'


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The program will substitute numbers in for the curly braces, and still print out all of the other text in the string:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If there are three sets of curly braces, the computer will expect three values to be listed in the format command. The first value given will replace the first curly brace.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sometimes we may want to print the same value twice. Or show them in a different order than how they were fed into the format function. See below that by placing a number in the curly braces, we can specify which parameter passed into the format function we want printed out. REMEMBER! Parameters are numbered starting at 0, so x is considered parameter 0.

In [11]:
x = 5
y = 66
z = 777
print("C - '{2}' A - '{0}' B - '{1}' C again - '{2}'".format(x, y, z))

C - '777' A - '5' B - '66' C again - '777'


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We can still specify formatting information after a colon. We can see below that the code will show the values right justified with a field width of four:

In [12]:
x = 5
y = 66
z = 777
print("C - '{2:4}' A - '{0:4}' B - '{1:4}' C again - '{2:4}'".format(x, y, z))

C - ' 777' A - '   5' B - '  66' C again - ' 777'


## Strings

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Let's look at how to format strings. The following list looks terrible.

In [16]:
from __future__ import print_function 
my_fruit = ["Apples","Oranges","Grapes","Pears"]
my_calories = [4, 300, 70, 30]
 
for i in range(4):
    print(my_fruit[i], "are", my_calories[i], "calories.")

Apples are 4 calories.
Oranges are 300 calories.
Grapes are 70 calories.
Pears are 30 calories.


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Now try it using the format command. Note how we can put additional text and more than one value into the same line. Remember text will be left justified and numbers are right justified.

In [17]:
from __future__ import print_function 
my_fruit = ["Apples", "Oranges", "Grapes", "Pears"]
my_calories = [4, 300, 70, 30]
 
for i in range(4):
    print("{:7} are {:3} calories.".format(my_fruit[i],my_calories[i]) )

Apples  are   4 calories.
Oranges are 300 calories.
Grapes  are  70 calories.
Pears   are  30 calories.


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;That's outstanding, and it looks the way we want it. But what if we didn't want the numbers right justified, and the text left justified? We can use < and > characters like the following example...although that looks dumb.

In [18]:
from __future__ import print_function 
my_fruit = ["Apples", "Oranges", "Grapes", "Pears"]
my_calories = [4, 300, 70, 30]
 
for i in range(4):
    print("{:>7} are {:<3} calories.".format(my_fruit[i],my_calories[i]) )

 Apples are 4   calories.
Oranges are 300 calories.
 Grapes are 70  calories.
  Pears are 30  calories.


## Leading Zeros

In [21]:
for hours in range(8,9):
    for minutes in range(6,12):
        print("Time {}:{}".format(hours, minutes))

Time 8:6
Time 8:7
Time 8:8
Time 8:9
Time 8:10
Time 8:11


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We need to use leading zeros for displaying numbers in clocks. Rather than specify a 2 for the field width, instead use 02. This will pad the field with zeros rather than spaces.

In [22]:
for hours in range(8,9):
    for minutes in range(6,12):
        print("Time {:02}:{:02}".format(hours, minutes))

Time 08:06
Time 08:07
Time 08:08
Time 08:09
Time 08:10
Time 08:11


## Floating Point Numbers

In [24]:
from __future__ import print_function 
x = 0.1
y = 123.456789
print("{:.1}  {:.1}".format(x,y) )
print("{:.2}  {:.2}".format(x,y) )
print("{:.3}  {:.3}".format(x,y) )
print("{:.4}  {:.4}".format(x,y) )
print("{:.5}  {:.5}".format(x,y) )
print("{:.6}  {:.6}".format(x,y) )
print()
print("{:.1f}  {:.1f}".format(x,y) )
print("{:.2f}  {:.2f}".format(x,y) )
print("{:.3f}  {:.3f}".format(x,y) )
print("{:.4f}  {:.4f}".format(x,y) )
print("{:.5f}  {:.5f}".format(x,y) )
print("{:.6f}  {:.6f}".format(x,y) )

0.1  1e+02
0.1  1.2e+02
0.1  1.23e+02
0.1  123.5
0.1  123.46
0.1  123.457

0.1  123.5
0.10  123.46
0.100  123.457
0.1000  123.4568
0.10000  123.45679
0.100000  123.456789


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A format of .2 means to display the number with two digits of precision. Unfortunately this means if we display the number 123 which has three significant numbers rather than rounding it we get the number in scientific notation: 1.2e+02.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A format of .2f (note the f) means to display the number with two digits after the decimal point. So the number 1 would display as 1.00 and the number 1.5555 would display as 1.56.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A program can also specify a field width character. The format 10.2f does not mean 10 digits before the decimal and two after. It means a total field width of 10. So there will be 7 digits before the decimal, the decimal which counts as one more, and 2 digits after.

In [26]:
from __future__ import print_function 
x = 0.1
y = 123.456789
print("'{:10.1}'  '{:10.1}'".format(x,y) )
print("'{:10.2}'  '{:10.2}'".format(x,y) )
print("'{:10.3}'  '{:10.3}'".format(x,y) )
print("'{:10.4}'  '{:10.4}'".format(x,y) )
print("'{:10.5}'  '{:10.5}'".format(x,y) )
print("'{:10.6}'  '{:10.6}'".format(x,y) )
print()
print("'{:10.1f}'  '{:10.1f}'".format(x,y) )
print("'{:10.2f}'  '{:10.2f}'".format(x,y) )
print("'{:10.3f}'  '{:10.3f}'".format(x,y) )
print("'{:10.4f}'  '{:10.4f}'".format(x,y) )
print("'{:10.5f}'  '{:10.5f}'".format(x,y) )
print("'{:10.6f}'  '{:10.6f}'".format(x,y) )

'       0.1'  '     1e+02'
'       0.1'  '   1.2e+02'
'       0.1'  '  1.23e+02'
'       0.1'  '     123.5'
'       0.1'  '    123.46'
'       0.1'  '   123.457'

'       0.1'  '     123.5'
'      0.10'  '    123.46'
'     0.100'  '   123.457'
'    0.1000'  '  123.4568'
'   0.10000'  ' 123.45679'
'  0.100000'  '123.456789'


## Printing Dollars and Cents

In [27]:
from __future__ import print_function 
cost1  = 3.07
tax1   = cost1 * 0.06
total1 = cost1 + tax1
 
print("Cost:  ${0:5.2f}".format(cost1) )
print("Tax:    {0:5.2f}".format(tax1) )
print("------------")
print("Total: ${0:5.2f}".format(total1) )

Cost:  $ 3.07
Tax:     0.18
------------
Total: $ 3.25


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remember! It would be easy to think that :5.2f would mean five digits, a decimal, followed by two digits. But it does not. It means a total field width of five, including the decimal and the two digits after.

<table width="100%" border="1" cellpadding="2" cellspacing="0">
<tr>
<td width="90" bgcolor="#00FF00" style="vertical-align:top"><img src="http://www.hermonswebsites.com/PyNotebooks/Python/sith.png" style="border:2px solid #021a40;"><center><b>
Jedi<br>Training<br>Task</b></center></td>
<td valign="top"><center><font size="+2"><b>Formatting Fibonacci</b></font></center><br>
<font size="+1">Write a function called fibonacci() that will print up to a maximum of the first 100 numbers in the Fibonacci sequence. If you don't know what the Fibonacci sequence is you should Google it. Pass the number into the function. The resulting fibonacci sequence must be right justified and contain commas.
</td></tr></table>

<table width="100%" cellpadding="2" cellspacing="2">
<tr>
<td style="border: 1px solid white;" width="100px">
<div>
<a href="http://nbviewer.ipython.org/github/CSP-UHS/Hermons_Book_of_Python/blob/master/Chapter06.ipynb"><img src="https://dl.dropboxusercontent.com/u/90219577/Hermon/CSE_Website/notebooks/leftarrow.png" height="50px" width="50px" align="left"></a>
</div>
</td>
<td style="border: 1px solid white;">
<div align="center">
<a href="http://nbviewer.ipython.org/github/CSP-UHS/Hermons_Book_of_Python/blob/master/Chapter00.ipynb"><h2>Table of Contents</h2></a>
</div>
</td>
<td style="border: 1px solid white;" width="100px">
<div>
<a href="http://nbviewer.ipython.org/github/CSP-UHS/Hermons_Book_of_Python/blob/master/Chapter08.ipynb"><img src="https://dl.dropboxusercontent.com/u/90219577/Hermon/CSE_Website/notebooks/rightarrow.png"  height="50px" width="50px"  align="right"></a>
</div>
</td>
</tr></table>